index.bs

```diff
@@ -3699,6 +3727,9 @@ Verifiers may use the following algorithm to verify an encoded {{CollectedClient
 1. Append [=CCDToString=](|challengeBase64|) to |expected|.
 1. Append 0x2c226f726967696e223a (`,"origin":`) to |expected|.
 1. Append [=CCDToString=](|origin|) to |expected|.
+1. If |topOrigin| is specified:
+    1. Append 0x2c22746f704f726967696e223a (`,"topOrigin":`) to |expected|.
+    1. Append [=CCDToString=](|topOrigin|) to |expected|.
```

Comment:
Again, this needs to go after `crossOrigin`.

Reply:
I also thought about whether this algorithm would be simpler if `topOrigin` is always present even when equal to `origin`, but I think that it ultimately doesn't matter since L2 browser implementations would still emit client data without `topOrigin` even if L3 always requires it.

Reply:
Apologies for the delay here - I've been pondering on this on and off over the week. I think there's an underlying question about whether the limited verification algorithm assumes a correct L3 implementation, or if it tries to 'catch' incorrect implementations.

Assuming that the underlying implementation is correct, then the 2x2 combination of inputs seems to be:

(a) crossOrigin=yes, topOrigin=null - verifier expected a crossOrigin case but does not care about the topOrigin
(b) crossOrigin=yes, topOrigin=foo.org - verifier expected a crossOrigin case and expect it to be from foo.org **if** that information is available (aka if the implementation is L3)
(c) crossOrigin=no, topOrigin=foo.org - invalid input (should specify this probably)
(d) crossOrigin=no, topOrigin=null - verifier expected a same-origin case

For that, this is what I have currently:

13. If _expected_ is not a prefix of _clientDataJSON_ then verification has failed
14. Let _remainingClientDataJSON_ be the result of removing the prefix _expected_ from _clientDataJSON_
15. If _topOrigin_ is specified and ",topOrigin:" is a prefix of _remainingClientDataJSON_ then:
    1. Let _expectedTopOrigin_ be an empty byte string
    2. Append ",topOrigin:" to _expectedTopOrigin_
    3. Append _topOrigin_ to _expectedTopOrigin_
    4. if _expectedTopOrigin_ is not a prefix of _remainingClientDataJSON_, fail
    5. let _remainingClientDataJSON_ be the result of removing the prefix _expectedTopOrigin_ from _clientDataJSON_
16. If _remainingClientDataJSON_ is not at least one byte long, fail
17. If the first byte of _remainingClientDataJSON_ is:
    - 0x7d => success
    - 0x2c => success
    - otherwise => fail

(a) is tested by step 13 alone, and then we skip step 15 because the verifier doesn't care
(b) is tested by step 13 and step 15 for L3 implementations, or just by step 13 for L2 implementations
(c) should be disallowed before the algorithm begins
(d) is tested by step 13 alone, and then we skip step 15 because any correct implementation will not have specified topOrigin


Does that make sense? My assumptions and/or algorithm may each be independently incorrect, but I'd like to validate my assumptions about what we want the verifier to be able to assume and to test first!

Reply:
That looks fair to me, I think. I'm generally in favour of strictness, so I reckon an explicit check of (c) is probably good. But I'd like for @agl to weigh in on this too.

Reply:
I would suggest keeping this algorithm simple and not worrying too much about it. There was a particular case that absolutely wouldn't accept parsing JSON and thus this exists, but I don't think the spec should put too much effort into handling every case without a suggestion that it would actually be useful for someone. Perhaps `topOrigin` could just be ignored in this algorithm?

Reply:
As per discussion in today's WebAuthn WG sync, I have removed any modification to the LVA, pending its potential deletion in the future.